extensions/amp-story/0.1/amp-story-auto-ads.js

```diff
@@ -29,42 +34,151 @@ export class AmpStoryAutoAds extends AMP.BaseElement {
         `<${TAG}> should be child of <amp-story>`);
-    this.ampStory_ = ampStory;
-  }
+
+    ampStoryElement.getImpl().then(impl => {
+      this.ampStory_ = impl;
+      this.navigationState_ = this.ampStory_.getNavigationState();
+      this.navigationState_.observe(this.handleStateChange_.bind(this));
+    });
+
+    // TODO(ccordry): move this chunk to layoutCallback
+    // TODO(ccordry): need to fake distance from page to force load
+    for (let i = 0; i < 2; i++) {
+      const mockPage = this.makeMockPage();
+      this.adElements_.push(mockPage);
+      ampStoryElement.appendChild(mockPage);
+    }
+  };
+
 
   /** @override */
   isLayoutSupported() {
     return true;
   }
 
+
   /** @override */
-  layoutCallback() {
-    this.ampStory.appendChild(this.makeMockPage());
-  }
+  // layoutCallback() {}
+
 
   // temporary to be replaced with real page later
   makeMockPage() {
     const ampStoryAdPage = document.createElement('amp-story-page');
-    ampStoryAdPage.id = 'i-amphtml-ad-page-1';
+    const id = this.adElements_.length + 1;
+    ampStoryAdPage.id = `i-amphtml-ad-page-${id}`;
+    ampStoryAdPage.setAttribute('advertisement', '');
     ampStoryAdPage./*OK*/innerHTML = `
       <amp-story-grid-layer template="vertical">
-        <h1>First Ad Page</h1>
-        <p>This is the first ad shown in this story.</p>
+        <h1>Ad Page #${id}</h1>
+        <p>This is ad #${id} shown in this story.</p>
       </amp-story-grid-layer>
       `;
     return ampStoryAdPage;
   }
+
+
+  /**
+   * Get array containing all pages of associated story
+   * @private
+   */
+  getPages_() {
+    if (!this.pages_) {
+      this.pages_ = this.ampStory_.getPages();
+    }
+    return this.pages_;
+  }
+
+
+  /**
+   * Get array containing all pages of associated story
+   * @private
+   */
+  handleStateChange_(stateChangeEvent) {
+    switch (stateChangeEvent.type) {
+      case StateChangeType.ACTIVE_PAGE:
+        const {pageIndex, pageId} = stateChangeEvent.value;
+        this.handleActivePageChange_(
+            dev().assertNumber(pageIndex),
+            dev().assertString(pageId));
+        break;
+    }
+  }
+
+
+  /**
+   * @param {number} unusedPageIndex
+   * @param {string} unusedPageId
+   * @private
+   */
+  handleActivePageChange_(unusedPageIndex, unusedPageId) {
+    this.interactions_++;
+    // temp before config in passed in
+    if (this.interactions_ > MIN_INTERVAL) {
+      this.placeNextAd_();
+      this.interactions_ = 0;
+    }
+  }
+
+
+  /**
+   * Place ad based on user config
+   * @private
+   */
+  placeNextAd_() {
+    // TODO(ccordry) make sure ad is loaded
+    const nextAdElement = this.adElements_[this.adIdPointer_];
+    const nextAdId = nextAdElement && nextAdElement.id;
+
+    if (!nextAdId) {
+      return;
+    }
+
+    const pages = this.getPages_();
+
+    // make methods public if necessary once finalized
+    const activePage = this.ampStory_.getActivePage();
```

Comment:
I feel like from here down, it may be better to just expose an `insertPage` function from `AmpStory`, and not leak its internals here

Reply:
This comes from a desire to have as little new code as possible in `amp-story` to make this new component work. I do see your point and could make potentially make things cleaner. We can chat about this tomorrow

Reply:
implemented insert page